### PR TITLE
api: changes in the api for imperative apps

### PR DESCRIPTION
**Pull request by raghavendra-talur**

### CODE REVIEW

File: api/v1alpha1/drplacementcontrol_types.go [hunk 1]
```diff
@@ -104,7 +119,15 @@ const (
 // DRPlacementControlSpec defines the desired state of DRPlacementControl
 type DRPlacementControlSpec struct {
 	// PlacementRef is the reference to the PlacementRule used by DRPC
-	PlacementRef v1.ObjectReference `json:"placementRef"`
+	PlacementRef v1.ObjectReference `json:"placementRef,omitempty"`
```
Comment (BenamarMk) [inline api/v1alpha1/drplacementcontrol_types.go:122]: This change makes it no longer a required field.  We need to make sure that we validate the DRPC CR early that the PlacementRef is provided when `OCMDiscovered` apps are not used. Or should we just leave it as `required` field but provide a no-op PlacementRef?

Comment (raghavendra-talur) [reply api/v1alpha1/drplacementcontrol_types.go:122]: Good question. I am leaning towards the first approach and therefore the changes in the drpc reconcile function. If I read the ramen config map early and also inspect the drpc spec, I will be able to determine whether the drpc is a OCM managed one or OCM discovered one. This way there both managed and discovered apps are treated the same. In the second approach, discovered apps feel like an afterthought.

File: api/v1alpha1/drplacementcontrol_types.go [hunk 2]
```diff
@@ -23,6 +23,21 @@ const (
 	ActionRelocate = DRAction("Relocate")
 )
 
+// DRPCType is the type of the DRPC based on the application handle
+type DRPCType string
+
+// These are the valid values for DRPCType
+const (
+	// OCMManaged is the DRPC type for the applications that are managed by OCM
+	OCMManaged = DRPCType("OCM Managed")
```
Comment (ShyamsundarR) [inline api/v1alpha1/drplacementcontrol_types.go:32]: Remove OCM from names, we can just call these `Managed`, 'Discovered' (or maybe Unmanaged)

File: api/v1alpha1/drplacementcontrol_types.go [hunk 3]
```diff
@@ -104,7 +119,14 @@ const (
 // DRPlacementControlSpec defines the desired state of DRPlacementControl
 type DRPlacementControlSpec struct {
 	// PlacementRef is the reference to the PlacementRule used by DRPC
-	PlacementRef v1.ObjectReference `json:"placementRef"`
+	PlacementRef v1.ObjectReference `json:"placementRef,omitempty"`
+
+	// PrimaryProtectedNamespace is the namespace where the application is running on the dr cluster.
+	// This field is mandatory when the DRPC is created for discovered or imperative apps.
+	PrimaryProtectedNamespace string `json:"primaryProtectedNamespace,omitempty"`
```
Comment (ShyamsundarR) [inline api/v1alpha1/drplacementcontrol_types.go:126]: Do we need 2 namespace fields, can we just use `EligibleNamespaces` as a single field and use it for required purposes?

The Recipe reference would be from one of the eligible namespaces

File: config/crd/bases/_.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+---
```
Comment (ShyamsundarR) [inline config/crd/bases/_.yaml:1]: This file gets generated for me as well for some time, unsure where the source for this is. It should not be committed (and the generation fixed as well)

File: api/v1alpha1/volumereplicationgroup_types.go [hunk 5]
```diff
@@ -185,6 +190,13 @@ type VolumeReplicationGroupSpec struct {
 	Action VRGAction `json:"action,omitempty"`
 	//+optional
 	KubeObjectProtection *KubeObjectProtectionSpec `json:"kubeObjectProtection,omitempty"`
+
+	// PrimaryProtectedNamespace is the namespace where the application is running on the dr cluster.
+	// This field is mandatory when the VRG is created for discovered or imperative apps.
+	PrimaryProtectedNamespace string `json:"primaryProtectedNamespace,omitempty"`
```
Comment (ShyamsundarR) [inline api/v1alpha1/volumereplicationgroup_types.go:196]: (same namespace comment as in DRPC)

**Comment by raghavendra-talur** [comment]
Abandon this PR as the request is to create a single PR with all the commits.

**Pull request closed without merging**